### PR TITLE
feat(middleware): add opa middleware

### PR DESCRIPTION
**Pull request by RichiCoder1**
# Description

Adds a new Open Policy Agent HTTP Middleware that allows for evaluating requests against rego snippets.

## Issue reference

https://github.com/dapr/dapr/issues/1617

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Created issue in the https://github.com/dapr/docs/ repo: https://github.com/dapr/docs/issues/786 (WIP)

### Example Usage

Adapted from https://www.openpolicyagent.org/docs/latest/http-api-authorization/#8-optional-use-json-web-tokens-to-communicate-policy-data
```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: authz
  namespace: default
spec:
  type: middleware.http.opa
  metadata:
  - name: rego
    value: |
        package http
        import input.request

        default allow = false

        # Allow users to get their own salaries.
        allow {
            some username
            request.method == "GET"
            path = ["finance", "salary", username]
            token.payload.user == username
        }

        # Allow managers to get their subordinate' salaries.
        allow {
            some username
            request.method == "GET"
            path = ["finance", "salary", username]
            token.payload.subordinates[_] == username
        }

        # Allow HR members to get anyone's salary.
        allow {
            request.method == "GET"
            path = ["finance", "salary", _]
            token.payload.hr == true
        }

        path = path {
            path := split(trim(input.path, "/"), "/")
        }

        token = {"payload": payload} {
            auth_header := request.headers["authorization"]
            [_, jwt] := split(auth_header, " ")
            [_, payload, _] := io.jwt.decode(input.token)
        }        
```

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: authz
  namespace: default
spec:
  type: middleware.http.opa
  metadata:
  - name: rego
    value: |
        package http
        import input.request

        default allow = true

        allow = {
            "allow": false, 
            "status_code": 403, 
            "additional_headers": { "www-authenticate": "Bearer realm=\"my_api\"" } 
        } {
            not token.payload.aud["my_api"]
        }

        token = {"payload": payload} {
            auth_header := request.headers["authorization"]
            [_, jwt] := split(auth_header, " ")
            [_, payload, _] := io.jwt.decode(input.token)
        }

```

### CODE REVIEW

**Comment by yaron2** [comment]
Can you add a small description of how this middle ware is used from the user's perspective?

/cc @mchmarney

**Comment by RichiCoder1** [comment]
Added some (admittedly contrived) examples to the description.

**Comment by RichiCoder1** [comment]
Thanks to someone else who pointed out an example, it looks like the rest of OPA is packaged up enough that it wouldn't be too hard to do a follow on that supports other OPA capabilities later too: https://github.com/open-policy-agent/example-api-authz-go/blob/master/internal/opa/opa.go

**Comment by yaron2** [comment]
@RichiCoder1 you marked this PR as a draft, which other changes are you expecting to push before this is ready for review?

**Comment by RichiCoder1** [comment]
> @RichiCoder1 you marked this PR as a draft, which other changes are you expecting to push before this is ready for review?

I think I just wanted to make sure everything was working and that the general idea was correct before I marked it read. Should be good now I think

**Comment by yaron2** [comment]
> > @RichiCoder1 you marked this PR as a draft, which other changes are you expecting to push before this is ready for review?
> 
> I think I just wanted to make sure everything was working and that the general idea was correct before I marked it ready. Should be good now I think
> 
> Edit: Will keep the `[WIP]` flag though until I get some confirmation I'm on the right path/people are on board with how this is being implemented as an MVP

Sounds good, thanks.

File: go.mod
Comment (yaron2) [inline go.mod:28]: Can you use `"github.com/stretchr/testify/assert"`? this is the assertion package we use through-out Dapr.

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: go.mod
Comment (RichiCoder1) [reply go.mod]: Hmm. My bad, must have been an autocomplete fail.

**Comment by yaron2** [comment]
@tcnghia can you please review this PR?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:37]:
```suggestion
// GetHandler returns the HTTP handler provided by the middleware
```

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:1]: Since we are already in package `opa`, can you please rename the file to just `middleware.go`?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:21]: Change to `NewMiddleWare`? That would remove the stutter in downstream call (i.e. `opa.NewMiddleWare()`)

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:17]: Can this be array instead of what appears to be some sort of delimited list?

I think that has the bonus of grouping the `strings.Split(...)` call on line #68 closer to the parsing logic.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:15]: Change to `middlewareMetadata`?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:105]: Can you please extract this into a var outside of this function, like `errRegoNoResult`

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:66]: Can you please document the meaning of the output `bool` here?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:94]: Please file tracking issues for these TODO as these are very hard to keep track and follow up later.

After that, we can replace `TODO` with `TODO(#issue)` or `TODO(<link-to-issue-comment>)`.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:104]:
```suggestion
	if err != nil {
		m.opaError(ctx, err)
		return false
	}
	
	if len(results) == 0 {
```

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:110]:
```suggestion
	}
	return m.handleRegoResult(ctx, results[0].Bindings["result"])
```

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:119]:
```suggestion
		if !allowed {
			ctx.Error(fasthttp.StatusMessage(fasthttp.StatusUnauthorized), fasthttp.StatusUnauthorized)
		}
		return allowed
```

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:123]: Can you please extract this err to a var outside of this func, e.g. `errInvalidRegoPolicyType`?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:135]: Can you please either link to rego documentation to explain what StatusCode -1 mean? Is there a constant with descriptive name that we can use? If not, can you please extract this to a constant with a descriptive name?

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:113]: Please add a short doc comment to mention what the returned `bool` means in case of `true` or `false`.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:145]: Please add a comment explaining what this means.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:157]:
```suggestion
	for key, value := range regoResult.AdditionalHeaders {
		ctx.Request.Header.Set(key, value)
	}
	if !regoResult.Allow && regoResult.StatusCode >= 100 {
		ctx.Response.SetStatusCode(regoResult.StatusCode)
	}
	return regoResult.Allow
```

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware.go:162]: Please extract `x-dapr-opa-error` into a const, or reuse if there is already one somewhere? Also, can you please add some comments explaining why "1"? Is that intended to be a boolean (in which case can we use "true" instead)?

File: middleware/http/opa/opa_middleware_test.go
Comment (tcnghia) [inline middleware/http/opa/opa_middleware_test.go:16]: These tests are very similar in structure. May be consider writing using table-driven test https://dave.cheney.net/2019/05/07/prefer-table-driven-tests and reuse much of the skeleton? Thanks!

**Review by tcnghia — CHANGES_REQUESTED** [review]

**Comment by RichiCoder1** [comment]
@tcnghia Thanks for all the feedback! Will dig in and address short. For at least some of it though, specifically the duplication in naming, I was mostly following what the other Http midlewares were doing. I'm fine (and actually prefer) removing the duplicate naming, but I wanted to make sure it was ok to break with the other middleware first.

File: middleware/http/opa/opa_middleware.go
Comment (RichiCoder1) [reply middleware/http/opa/opa_middleware.go:17]: Is that something that the middleware metadata will allow? If so, I'll make this swap, I'd much rather have it be an array.

File: middleware/http/opa/opa_middleware.go
Comment (RichiCoder1) [reply middleware/http/opa/opa_middleware.go:162]: Will do on extracting the constant name. The value being `1` is an http-ism for truthyness. I'm cool doing `true` too.

File: middleware/http/opa/opa_middleware_test.go
Comment (RichiCoder1) [reply middleware/http/opa/opa_middleware_test.go:16]: Oooo, that's neat. I'll get the tests converted, certaintly a lot of duplication hapening.

File: middleware/http/opa/opa_middleware.go
Comment (RichiCoder1) [reply middleware/http/opa/opa_middleware.go:94]: Wanted to make sure the `opa` middleware was a go before I extracted these out. If consensus is yes on adding the middleware, I can get these marshaled to issues.

File: middleware/http/opa/opa_middleware.go
Comment (yaron2) [reply middleware/http/opa/opa_middleware.go:17]: Metadata for components only supports key/Val's, so this does in fact need to be a delimited string..

File: middleware/http/opa/opa_middleware.go
Comment (yaron2) [reply middleware/http/opa/opa_middleware.go:94]: This was discussed in a maintainers meeting and was decided to be included, provided the PR is approved and that there's a docs update for the middleware :)

Comment (RichiCoder1) [reply middleware/http/opa/opa_middleware.go:94]: @yaron2 Cool beans! I'll spin out these tickets and start a docs pr then

**Comment by tcnghia** [comment]
@RichiCoder1 thanks for addressing the feedback! I will leave it to @yaron2 to comment about the overall theme.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [reply middleware/http/opa/opa_middleware.go:17]: Even if the metadata value is a string, when parsing it we can use array in this intermediate structure (`opaMiddlewareMetadata`) so that no consumer of such data structure need to do the string split.

File: middleware/http/opa/opa_middleware.go
Comment (tcnghia) [reply middleware/http/opa/opa_middleware.go]: I see, thanks for explaining. The two blocks looks so similar and I overlooked the difference. Thanks for adding a comment.

File: middleware/http/opa/middleware.go
Comment (tcnghia) [inline middleware/http/opa/middleware.go:41]: you can call errors.New() here and avoid calling when use. I don't think errors.New() decorate things with stacktrace, so there is no need to create new one every time. 

If you do so, it is customary to name an error like `errFoo`.

File: middleware/http/opa/middleware.go
Comment (tcnghia) [inline middleware/http/opa/middleware.go:160]: This reads really well now, thanks a lot!

**Review by tcnghia — CHANGES_REQUESTED** [review]
Looking good! One more comment.

File: middleware/http/opa/middleware.go
Comment (RichiCoder1) [reply middleware/http/opa/middleware.go:41]: Awesome! I wasn't sure, will fix that up now.

Comment (RichiCoder1) [reply middleware/http/opa/middleware.go]: Fixed up

**Comment by tcnghia** [comment]
@RichiCoder1 thanks a lot for adding more tests. awesome.

@yaron2 I don't have anything else to add, this looks good to me.

File: middleware/http/opa/opa_middleware.go
Comment (yaron2) [reply middleware/http/opa/opa_middleware.go]: That's possible, but basically there are no other consumers of the metadata struct, it's internal to the Middleware which gets registered in the core Dapr runtime with an interface.

**Review by yaron2 — APPROVED** [review]
/lgtm

**Comment by yaron2** [comment]
Thanks @tcnghia and @RichiCoder1.

**Comment by RichiCoder1** [comment]
🥳 Whoop! Happy to have contributed. Is there anything else I need to do to get this integrated aside from docs?

**Comment by yaron2** [comment]
> 🥳 Whoop! Happy to have contributed. Is there anything else I need to do to get this integrated aside from docs?

Yep. Please open a PR in the main Dapr repo to register the middleware. It's pretty straightforward, see here: https://github.com/dapr/dapr/blob/1e3b7aa3202e268759bce6cf9b30b6d95471ab8c/cmd/daprd/main.go#L387

To get the latest version of contrib with the OPA middleware, run `go get -u github.com/dapr/components-contrib@master` and then make the changes in `main.go`.

**Pull request merged**
